pulser/sequence.py

```diff
@@ -12,32 +12,60 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from __future__ import annotations
+
 from collections import namedtuple
-from collections.abc import Iterable
+from collections.abc import Callable, Generator, Iterable, Set
 import copy
 from functools import wraps
 from itertools import chain
 import json
-import warnings
-
+from numpy.typing import ArrayLike
 import numpy as np
+from sys import version_info
+from typing import Any, cast, NamedTuple, Optional, Tuple, Union
+import warnings
```

Comment:
Just move these to the bottom of the first block and bring `numpy.typing` here, the rest is okay.

Reply:
Also, put `numpy` imports one the second block, as it is a third-party library